src/main/java/org/gridsuite/modification/server/service/NetworkModificationService.java

```diff
@@ -13,43 +16,82 @@
-import org.springframework.beans.factory.annotation.Autowired;
-import org.springframework.context.annotation.ComponentScan;
 import org.springframework.stereotype.Service;
+import reactor.core.publisher.Flux;
 import reactor.core.publisher.Mono;
 import reactor.core.scheduler.Schedulers;
 
-import java.util.Set;
-import java.util.UUID;
-
 import static org.gridsuite.modification.server.NetworkModificationException.Type.*;
 
 /**
  * @author Franck Lecuyer <franck.lecuyer at rte-france.com>
+ * @author Slimane Amar <slimane.amar at rte-france.com>
  */
-@ComponentScan(basePackageClasses = {NetworkStoreService.class})
 @Service
-class NetworkModificationService {
+public class NetworkModificationService {
 
     private static final Logger LOGGER = LoggerFactory.getLogger(NetworkModificationService.class);
 
-    @Autowired
-    private NetworkStoreService networkStoreService;
+    private final NetworkStoreService networkStoreService;
+
+    private final NetworkModificationRepository modificationRepository;
+
+    public NetworkModificationService(NetworkStoreService networkStoreService, NetworkModificationRepository modificationRepository) {
+        this.networkStoreService = networkStoreService;
+        this.modificationRepository = modificationRepository;
+    }
+
+    public Flux<ElementaryModificationInfos> applyGroovyScript(UUID networkUuid, String groovyScript) {
+        return assertGroovyScriptNotEmpty(groovyScript).thenMany(
+                getNetwork(networkUuid).flatMapIterable(network -> doModification(network, networkUuid, () -> {
+                    CompilerConfiguration conf = new CompilerConfiguration();
+                    Binding binding = new Binding();
+                    binding.setProperty("network", network);
+                    GroovyShell shell = new GroovyShell(binding, conf);
+                    shell.evaluate(groovyScript);
+                }, GROOVY_SCRIPT_ERROR))
+        );
+    }
+
+    public Flux<ElementaryModificationInfos> changeSwitchState(UUID networkUuid, String switchId, boolean open) {
+        return getNetwork(networkUuid)
+                .filter(network -> network.getSwitch(switchId) != null)
+                .switchIfEmpty(Mono.error(new NetworkModificationException(SWITCH_NOT_FOUND, switchId)))
+                .filter(network -> network.getSwitch(switchId).isOpen() != open)
+                .flatMapIterable(network -> doModification(network, networkUuid, () -> network.getSwitch(switchId).setOpen(open)));
+    }
+
+    public Flux<UUID> getModificationGroups() {
+        return Flux.fromIterable(modificationRepository.getModificationGroupsUuids());
+    }
 
-    private Set<String> doModification(Network network, Runnable modification) {
-        return doModification(network, modification, MODIFICATION_ERROR);
+    public Flux<ModificationInfos> getModifications(UUID groupUuid) {
+        return Flux.fromIterable(modificationRepository.getModifications(groupUuid));
```

Comment:
do all db access during subscribe, not during mono creation

Reply:
Done